### PR TITLE
feat: session validation callbacks

### PR DESCRIPTION
**Pull request by alisher-aituarov**
## Summary of change

Provide user with ability to control redirections by specifying callbacks (onFailure(Success)Redirection) 

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [x] Changelog has been updated
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [x] Had run `npm run build-pretty`
-   [x] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If added a new recipe interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [ ] If I added a new recipe, I also added the recipe entry point into the `size-limit` section of `package.json` with the size limit set to the current size rounded up.
-   [ ] If I added a new recipe, I also added the recipe entry point into the `rollup.config.mjs`

## Remaining TODOs for this PR

-   [ ] Remove the `userContext={{}}` part in SessionAuth with-update-email-post-verification example app, and then test that the tests for the example app work on Rishabh's laptop. Specifically, without the `userContext={{}}` on the older version of the sdk, after email api returns a `202`, there was no redirection (to the email verification screen) happening in the automated tests. Adding `userContext={{}}` fixed that, but we need to remove it.

### CODE REVIEW

File: lib/ts/recipe/session/recipe.ts [hunk 1]
```diff
@@ -158,21 +153,53 @@ export default class Session extends RecipeModule<unknown, unknown, unknown, any
...
+                for (const claim of invalidClaims) {
+                    const validator = globalValidators.find(({ id }) => id === claim.validatorId);
+                    if (validator?.onFailureRedirection) {
+                        callbacks.push(validator?.onFailureRedirection?.());
+                    }
+                }
+                return (await Promise.all(callbacks)).filter(Boolean)[0];
+            };
+
+            const redirectPath = await getRedirectionPath();
```
Comment (porcellus) [inline lib/ts/recipe/session/recipe.ts:182]: 1. I don't see a reason why this is extracted into a function
2. We shouldn't call all of them at the same time: 
  a/ it might be misleading to users (if they return a value from it they could expect the redirection to happen)
  b/ it could be called in non-deterministic order, which may cause issues when testing in the future
  c/ there is no reason to do it: we only want the first one, and they are almost always sync functions anyway.

I think both this and the success part could be simplified into just a for loop (both could be done with reduce, but it's kind of awkward with promises)

File: lib/ts/recipe/session/sessionAuth.tsx [hunk 2]
```diff
@@ -125,15 +125,14 @@ const SessionAuth: React.FC<PropsWithChildren<SessionAuthProps>> = ({ children,
                 userId: "",
             };
         }
-        // TODO: basing redirection on userContext could break in certain edge-cases involving async validators
-        const invalidClaimRedirectToPath = popInvalidClaimRedirectPathFromContext(userContext);
+        // // TODO: basing redirection on userContext could break in certain edge-cases involving async validators
+        // const invalidClaimRedirectToPath = popInvalidClaimRedirectPathFromContext(userContext);
```
Comment (porcellus) [inline lib/ts/recipe/session/sessionAuth.tsx:129]: These can be removed

File: lib/ts/recipe/session/sessionAuth.tsx [hunk 3]
```diff
@@ -161,8 +160,42 @@ const SessionAuth: React.FC<PropsWithChildren<SessionAuthProps>> = ({ children,
         }
     }, []);
 
+    const getRedirectPathFromInvalidClaimsOrRedirect = async (claims: ClaimValidationError[]) => {
+        if (!claims.length) {
+            return;
+        }
+        const globalValidators: SessionClaimValidator[] = getGlobalClaimValidators(props.overrideGlobalClaimValidators);
```
Comment (porcellus) [inline lib/ts/recipe/session/sessionAuth.tsx:167]: Do we need to specify the type here?

File: lib/ts/recipe/session/sessionAuth.tsx [hunk 4]
```diff
@@ -161,8 +160,42 @@ const SessionAuth: React.FC<PropsWithChildren<SessionAuthProps>> = ({ children,
...
+
+        // we try to find claim validator among failed validators with onFailure cb that returns string
+        const getRedirectionPath = async () => {
```
Comment (porcellus) [inline lib/ts/recipe/session/sessionAuth.tsx:170]: The comments of `validateGlobalClaimsAndHandleSuccessRedirection` apply here as well.
Also, please check if you could extract this into a util.

File: lib/ts/types.ts [hunk 5]
```diff
@@ -21,10 +21,20 @@ import { NormalisedConfig as NormalisedRecipeModuleConfig } from "./recipe/recip
...
+import { SessionClaimValidator as SessionClaimValidatorWebJS } from "supertokens-web-js/recipe/session";
+
+export type GetRedirectionURLContext =
+    | {
+          action: "TO_AUTH" | "SESSION_VERIFICATION_FAILURE";
+          showSignIn?: boolean;
+      }
+    | {
+          action: "SESSION_VERIFICATION_FAILURE";
+      };
```
Comment (porcellus) [inline lib/ts/types.ts:33]:
```suggestion
    | {
          action: "TO_AUTH";
          showSignIn?: boolean;
      }
    | {
          action: "SESSION_CLAIM_VERIFICATION_FAILURE";
      };
```

File: lib/ts/recipe/session/index.ts [hunk 6]
```diff
@@ -14,21 +14,18 @@
...
-} from "supertokens-web-js/recipe/session";
+import { ClaimValidationResult, RecipeInterface, SessionClaim } from "supertokens-web-js/recipe/session";
 import SessionAuthWrapper from "./sessionAuth";
 import useSessionContextFunc from "./useSessionContext";
 import { useClaimValue as useClaimValueFunc } from "./useClaimValue";
 import { InputType, SessionContextType } from "./types";
 import SessionContext from "./sessionContext";
 import { getNormalisedUserContext } from "../../utils";
-import { ClaimValidationError, SessionClaimValidator } from "supertokens-website";
+import { ClaimValidationError } from "supertokens-website";
```
Comment (porcellus) [inline lib/ts/recipe/session/index.ts:24]: This shouldn't import from supertokens-website

File: lib/ts/recipe/session/sessionAuth.tsx [hunk 7]
```diff
@@ -161,8 +160,42 @@ const SessionAuth: React.FC<PropsWithChildren<SessionAuthProps>> = ({ children,
         }
     }, []);
 
+    const getRedirectPathFromInvalidClaimsOrRedirect = async (claims: ClaimValidationError[]) => {
```
Comment (porcellus) [inline lib/ts/recipe/session/sessionAuth.tsx:163]: This should use `useCallback` and be added to the dependency arrays where called.

File: lib/ts/recipe/session/sessionAuth.tsx [hunk 8]
```diff
@@ -171,17 +204,16 @@ const SessionAuth: React.FC<PropsWithChildren<SessionAuthProps>> = ({ children,
...
-                } else if (toSetContext.invalidClaimRedirectToPath !== undefined) {
-                    await SuperTokens.getInstanceOrThrow().redirectToUrl(
-                        toSetContext.invalidClaimRedirectToPath,
-                        history
-                    );
+                }
+                // since we have claimID that means we can safely get redirection string and then can we do the redirection.
+                const redirectPath = await getRedirectPathFromInvalidClaimsOrRedirect(toSetContext.invalidClaims);
+
+                if (redirectPath !== undefined) {
```
Comment (porcellus) [inline lib/ts/recipe/session/sessionAuth.tsx:211]: Why do we do redirection both inside the function and outside

File: lib/ts/recipe/session/sessionAuth.tsx [hunk 9]
```diff
@@ -207,7 +239,7 @@ const SessionAuth: React.FC<PropsWithChildren<SessionAuthProps>> = ({ children,
                     });
                     setContext({ ...event.sessionContext, loading: false, invalidClaims });
 
-                    const redirectPath = popInvalidClaimRedirectPathFromContext(userContext);
+                    const redirectPath = await getRedirectPathFromInvalidClaimsOrRedirect(invalidClaims);
```
Comment (porcellus) [inline lib/ts/recipe/session/sessionAuth.tsx:242]: This would sometimes redirect even if doRedirection is false

File: lib/ts/recipe/session/recipe.ts [hunk 10]
```diff
@@ -158,21 +153,53 @@ export default class Session extends RecipeModule<unknown, unknown, unknown, any
...
+
+            // if none has onFailure that returns string we fallback to default
```
Comment (porcellus) [inline lib/ts/recipe/session/recipe.ts:184]: If onFailure is defined but returns undefined we do not need to redirect to the access denied page. 
In that case, we can continue with the onSuccess redirections (for the non-failing claims ofc).

**Review by porcellus — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/session/index.ts (hunk 6 above)
Comment (porcellus) [reply lib/ts/recipe/session/index.ts:24]: Please merge the master branch into this.

File: lib/ts/recipe/session/sessionAuth.tsx (hunk 3 above)
Comment (alisher-aituarov) [reply lib/ts/recipe/session/sessionAuth.tsx]: yes, web-js | website does not know about callbacks

**Review by porcellus — CHANGES_REQUESTED** [review]

File: lib/ts/types.ts [hunk 11]
```diff
@@ -18,12 +18,26 @@ import NormalisedURLDomain from "supertokens-web-js/utils/normalisedURLDomain";
...
+    action: "TO_AUTH" | "SESSION_VERIFICATION_FAILURE";
+    showSignIn?: boolean;
+};
+
+export type ValidationSuccessCallback =
+    | (({ userContext }: { userContext: any }) => Promise<string | undefined> | string)
+    | undefined;
+
+export type ValidationFailureCallback =
+    | (({ userContext, reason }: { userContext: any; reason: any }) => Promise<string | undefined> | string)
```
[the remark this answers is not in the document]
Comment (porcellus) [reply lib/ts/types.ts:35]: plus it should be able to return raw undefined as well.

File: lib/ts/utils.ts [hunk 12]
```diff
@@ -453,22 +455,56 @@ export const useOnMountAPICall = <T>(
...
+        accessForbidden: redirects.length === 0,
+        redirectPath: redirects.filter(Boolean)[0],
+    };
+};
 
-export function popInvalidClaimRedirectPathFromContext(userContext: any) {
-    const res = userContext["_default"]?.invalidClaimRedirectPath;
-    if (res !== undefined) {
-        delete userContext["_default"].invalidClaimRedirectPath;
+export const getSuccessRedirectionPath = async ({
```
[the remark this answers is not in the document]
Comment (alisher-aituarov) [reply lib/ts/utils.ts:492]: i added it to function parameters but there is no overrideGlobalClaimValidators in place where this function is used, validateClaims is not accepting overrideGlobalClaimValidators as well there (session/recipe.tsx)

File: lib/ts/types.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (alisher-aituarov) [reply lib/ts/types.ts:35]: export declare type ClaimValidationError = {
    validatorId: string;
    reason?: any;
};

File: lib/ts/recipe/passwordless/recipe.tsx [hunk 13]
```diff
@@ -75,14 +75,14 @@ export default class Passwordless extends AuthRecipe<
             const normalisedFullPath = this.config.appInfo.websiteBasePath.appendPath(new NormalisedURLPath("/"));
             features[normalisedFullPath.getAsStringDangerous()] = {
                 matches: matchRecipeIdUsingQueryParams(this.config.recipeId),
-                component: (props) => this.getFeatureComponent("signInUp", props),
+                component: (props) => this.getFeatureComponent("signInUp", props as any),
```
Comment (porcellus) [inline lib/ts/recipe/passwordless/recipe.tsx:78]: Why are these changes required?

File: lib/ts/recipe/session/sessionAuth.tsx [hunk 14]
```diff
@@ -207,9 +222,26 @@ const SessionAuth: React.FC<PropsWithChildren<SessionAuthProps>> = ({ children,
                     });
                     setContext({ ...event.sessionContext, loading: false, invalidClaims });
 
-                    const redirectPath = popInvalidClaimRedirectPathFromContext(userContext);
-                    if (props.doRedirection !== false && redirectPath) {
-                        await SuperTokens.getInstanceOrThrow().redirectToUrl(redirectPath, history);
+                    const failureRedirectInfo = await getFailureRedirectionInfo({
```
Comment (porcellus) [inline lib/ts/recipe/session/sessionAuth.tsx:225]: We don't even need to bother getting these if `doRedirection` is false

File: lib/ts/recipe/session/sessionAuth.tsx [hunk 15]
```diff
@@ -171,16 +166,36 @@ const SessionAuth: React.FC<PropsWithChildren<SessionAuthProps>> = ({ children,
...
+                }
+
+                const failureRedirectInfo = await getFailureRedirectionInfo({
+                    invalidClaims: toSetContext.invalidClaims,
+                    overrideGlobalClaimValidators: props.overrideGlobalClaimValidators,
+                    userContext,
+                });
+                if (failureRedirectInfo.redirectPath !== undefined) {
+                    await SuperTokens.getInstanceOrThrow().redirectToUrl(failureRedirectInfo.redirectPath, history);
+                    return;
```
Comment (porcellus) [inline lib/ts/recipe/session/sessionAuth.tsx:178]:
```suggestion
                    return await SuperTokens.getInstanceOrThrow().redirectToUrl(failureRedirectInfo.redirectPath, history);
```

File: lib/ts/types.ts (hunk 5 above)
Comment (porcellus) [reply lib/ts/types.ts]: This still didn't happen.

File: lib/ts/utils.ts [hunk 16]
```diff
@@ -453,22 +455,68 @@ export const useOnMountAPICall = <T>(
...
+            const redirectPath = await validator.onFailureRedirection({ reason: claim.reason, userContext });
+            redirects.push(redirectPath);
+            if (redirectPath !== undefined) {
+                break;
+            }
+        }
     }
-}
+    return {
+        accessForbidden: invalidClaims.length > 0 && redirects.length === 0,
```
Comment (porcellus) [inline lib/ts/utils.ts:490]: Using this, if any claim onFailureRedirection returns undefined, we don't redirect to the access denied page.
This is not correct.
If:
1. The app uses a RoleClaim that redirects to the access denied page by default
2. The app later adds a claim that shouldn't redirect (so it returns undefined both redirection callbacks)

Then:
- The RoleClaim redirection no longer happens.

Comment (porcellus) [reply lib/ts/utils.ts:490]: The priority of redirections should be:
1. A failed claim returns a string.
2. A failed claim doesn't have `onFailureRedirection` defined -> access denied
3. A valid claim has onSuccessRedirection, and it returns a string
4. Default behaviour (success event or no redirection, depending on the location)

File: lib/ts/utils.ts [hunk 17]
```diff
@@ -453,22 +455,68 @@ export const useOnMountAPICall = <T>(
...
+        redirectPath: redirects.filter(Boolean)[0],
```
Comment (porcellus) [inline lib/ts/utils.ts:491]: This will return undefined if the `onFailureRedirection` returns an empty string.

File: lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+    function AccessDeniedThemeScreen(props) {
+        const t = useTranslation();
+        const hasFont = hasFontDefined(props.config.rootStyle);
+
+        return (
+            <ThemeBase
+                loadDefaultFont={!hasFont}
+                userStyles={[props.config.rootStyle, props.config.accessDeniedScreen.style]}>
+                <Fragment>
+                    <div data-supertokens="accessDeniedTitle">{t("ACCESS_DENIED")}</div>
```
Comment (porcellus) [inline lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx:20]: We already have title styles, please use those (check the `INVALID` state of the `veryEmailLinkClicked` theme component.

**Review by porcellus — CHANGES_REQUESTED** [review]

File: lib/ts/utils.ts (hunk 17 above)
Comment (alisher-aituarov) [reply lib/ts/utils.ts:491]: yes which would mean that access is forbidden and we dont look whats in redirectPath

File: lib/ts/recipe/passwordless/recipe.tsx (hunk 13 above)
Comment (alisher-aituarov) [reply lib/ts/recipe/passwordless/recipe.tsx:78]: just to make typescript stop yelling, file is always marked as "with errors" without. these changes

File: lib/ts/utils.ts (hunk 17 above)
Comment (porcellus) [reply lib/ts/utils.ts:491]: I don't think that's true. If the first claim returns an empty string then we return: 
`{ accessForbidden: false, redirectPath: undefined}`
This is not the right behaviour.

File: lib/ts/recipe/session/recipe.tsx [hunk 19]
```diff
@@ -161,18 +167,43 @@ export default class Session extends RecipeModule<unknown, unknown, unknown, any
...
-            return SuperTokens.getInstanceOrThrow().redirectToUrl(invalidClaimRedirectPath, history);
+
+            // we try to find claim validator among failed validators with onFailure cb that returns string
+            const failureRedirectInfo = await getFailureRedirectionInfo({
+                invalidClaims,
+                userContext,
+            });
+
+            if (failureRedirectInfo.accessForbidden) {
+                const failedClaim = JSON.stringify(invalidClaims[0]);
```
[the remark this answers is not in the document]
Comment (alisher-aituarov) [reply lib/ts/recipe/session/recipe.tsx:186]: it looks like i need to do the Promise.all thing here after all, because i am not able to save and return failedClaim in loop with`await`

File: lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+                    <div data-supertokens="accessDeniedTitle">
+                        <div>Claim ID: {props.denialInfo?.validatorId}</div>
```
[the remark this answers is not in the document]
Comment (alisher-aituarov) [reply lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx:22]: i guess we need condition like `props.denialInfo && <jsx/>`, do we ?

File: examples/with-emailverification-then-password-thirdpartyemailpassword/src/realPasswordClaim.js [hunk 21]
```diff
@@ -1,8 +1,9 @@
-import { BooleanClaim } from "supertokens-web-js/recipe/session";
+import { BooleanClaim } from "supertokens-auth-react/recipe/session";
 
 export const RealPasswordClaim = new BooleanClaim({
```
[the remark this answers is not in the document]
Comment (alisher-aituarov) [reply examples/with-emailverification-then-password-thirdpartyemailpassword/src/realPasswordClaim.js:3]: but not all of them right ?

File: lib/ts/utils.ts (hunk 16 above)
Comment (alisher-aituarov) [reply lib/ts/utils.ts:490]: - Using this, if any claim onFailureRedirection returns undefined, we don't redirect to the access denied page.
This is not correct. 
i remember we discussed that if user defines onFailureRedirection and makes it return undefined that means he does not want it to redirect to access denied

File: examples/with-cli-login/src/App.tsx [hunk 22]
```diff
@@ -36,11 +36,6 @@ SuperTokens.init({
                     ThirdPartyEmailPassword.Apple.init(),
                 ],
             },
-            getRedirectionURL: async function (context) {
```
Comment (porcellus) [inline examples/with-cli-login/src/App.tsx:39]: Did you test this? I don't think this should be removed.

File: lib/ts/utils.ts [hunk 23]
```diff
@@ -453,22 +455,68 @@ export const useOnMountAPICall = <T>(
...
-    if (res !== undefined) {
-        delete userContext["_default"].invalidClaimRedirectPath;
+export const getSuccessRedirectionPath = async ({
+    invalidClaims,
+    overrideGlobalClaimValidators,
+    userContext,
+}: {
+    invalidClaims: ClaimValidationError[];
+    overrideGlobalClaimValidators?:
+        | ((globalClaimValidators: SessionClaimValidator[], userContext: any) => SessionClaimValidator[])
```
[the remark this answers is not in the document]
Comment (porcellus) [reply lib/ts/utils.ts]: This didn't change

File: lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx (hunk 20 above)
[the remark this answers is not in the document]
Comment (porcellus) [reply lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx:24]: I'm not sure what you mean, I meant could you paste a screenshot of this screen if one of the role claims fail?

File: lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+import type { AccessDeniedThemeProps } from "../../../types";
+
+export const AccessDeniedTheme = withOverride<React.FC<AccessDeniedThemeProps>>(
+    "SessionAccessDenied",
+    function AccessDeniedThemeScreen(props) {
+        const t = useTranslation();
+        const hasFont = hasFontDefined(props.config.rootStyle);
+
+        return (
+            <ThemeBase
```
Comment (porcellus) [inline lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx:18]: The ThemeBase shouldn't be inside the overrideable component.

File: examples/with-emailverification-then-password-thirdpartyemailpassword/src/realPasswordClaim.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (porcellus) [reply examples/with-emailverification-then-password-thirdpartyemailpassword/src/realPasswordClaim.js:3]: Well, make sure they still work as expected.

**Review by porcellus — CHANGES_REQUESTED** [review]
Please re-check the comments I've left unresolved

File: lib/ts/recipe/session/index.ts [hunk 25]
```diff
@@ -109,6 +106,8 @@ export default class SessionAPIWrapper {
             userContext: getNormalisedUserContext(input?.userContext),
         });
     }
+
+    static ComponentsOverrideProvider = RecipeComponentsOverrideContextProvider;
```
Comment (porcellus) [inline lib/ts/recipe/session/index.ts:110]: We also need to export the components.

File: test/unit/recipe/session/utils.unit.test.ts [hunk 26]
```diff
@@ -0,0 +1,66 @@
+import { getFailureRedirectionInfo } from "../../../../lib/ts/recipe/session/utils";
+import * as utils from "supertokens-web-js/utils";
+
+beforeEach(() => {
+    jest.spyOn(utils, "getGlobalClaimValidators").mockReturnValue([
+        {
+            id: "id-3",
```
Comment (porcellus) [inline test/unit/recipe/session/utils.unit.test.ts:7]: These tests would be easier to read if the ids were a bit more descriptive - e.g.: `no-onFailureRedirection`, `onFailureRedirection-returns-undefined`, etc.

File: test/unit/recipe/session/utils.unit.test.ts [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+            onFailureRedirection: () => undefined,
+        },
+        {
+            id: "id-1",
+            onFailureRedirection: () => "/id-1",
+        },
+    ] as any[]);
+});
+
+describe(`${getFailureRedirectionInfo.name} functionality`, () => {
```
Comment (porcellus) [inline test/unit/recipe/session/utils.unit.test.ts:21]: Add a test-case that shows that the order of redirection depends on `getGlobalClaimValidators`: two claims should return a string, the one that comes first in `getGlobalClaimValidators` should be used.

File: test/unit/recipe/session/sessionAuth.test.tsx [hunk 28]
```diff
@@ -480,7 +484,10 @@ describe("SessionAuth", () => {
             });
 
             const result = render(
-                <SessionAuth>
+                <SessionAuth
+                    overrideGlobalClaimValidators={() =>
```
Comment (porcellus) [inline test/unit/recipe/session/sessionAuth.test.tsx:488]: Why is the `overrideGlobalClaimValidators` needed?

File: lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx (hunk 20 above)
[the remark this answers is not in the document]
Comment (porcellus) [reply lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx]: I still don't see the screenshot...

File: lib/ts/recipe/session/sessionAuth.tsx (hunk 14 above)
Comment (porcellus) [reply lib/ts/recipe/session/sessionAuth.tsx:231]: I don't think this happened.

File: lib/ts/utils.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (porcellus) [reply lib/ts/utils.ts]: And it still uses both ? and | undefined for `overrideGlobalClaimValidators`.

**Review by porcellus — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx [hunk 29]
```diff
@@ -0,0 +1,63 @@
...
+    const userContext = useUserContext();
+    const t = useTranslation();
+    const history = SuperTokens.getReactRouterDomWithCustomHistory()?.useHistoryCustom();
+
+    const onLogout = async () => {
+        try {
+            await Session.getInstanceOrThrow().signOut({ userContext });
+        } catch (error) {
+        } finally {
+            history.navigate("/auth");
```
Comment (porcellus) [inline lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx:28]: This should use redirectToAuth instead.

File: lib/ts/recipe/session/index.ts [hunk 30]
```diff
@@ -108,6 +109,9 @@ export default class SessionAPIWrapper {
     }
 
     static ComponentsOverrideProvider = RecipeComponentsOverrideContextProvider;
+
+    static AccessDenied = (prop: PropsWithChildren<{ redirectOnSessionExists?: boolean; userContext?: any }> = {}) =>
```
Comment (porcellus) [inline lib/ts/recipe/session/index.ts:113]: The theme component should also be exported.

File: lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx [hunk 31]
```diff
@@ -1,28 +1,50 @@
...
+
+    const onBackButtonClicked = () => {
+        return SuperTokens.getInstanceOrThrow().redirectToAuth({
+            show: "signin",
+            history,
+            redirectBack: false,
+        });
+    };
 
     return (
```
Comment (porcellus) [inline lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx:40]: Please post a screenshot of this screen (into this thread or on the PR itself)

File: lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx [hunk 32]
```diff
@@ -1,28 +1,50 @@
...
+        }
+    };
+
+    const onBackButtonClicked = () => {
+        return SuperTokens.getInstanceOrThrow().redirectToAuth({
+            show: "signin",
+            history,
+            redirectBack: false,
+        });
+    };
```
Comment (porcellus) [inline lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx:38]: This isn't really a back button. What's the best we can do to emulate the "go back" button of the browser? Or would that just get us to the page that redirected us to this page in case it was a validation failure in `SessionAuth`?

**Review by porcellus — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/session/components/themes/styles.css [hunk 33]
```diff
@@ -0,0 +1,51 @@
...
+    background-color: #000000;
+    border-radius: 10px;
+    cursor: pointer;
+    align-items: center;
+    justify-content: center;
+    color: #ffffff;
+    font-weight: bold;
+}
+
+[data-supertokens~="backButtonCommon"] {
```
Comment (porcellus) [inline lib/ts/recipe/session/components/themes/styles.css:41]: Some of these styles are not used at all

File: lib/ts/recipe/session/components/themes/styles.css [hunk 34]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+[data-supertokens~="buttons"] {
+    display: flex;
+    gap: 20px;
+    align-items: center;
+    justify-content: center;
+}
+
+[data-supertokens~="logoutButton"] {
```
Comment (porcellus) [inline lib/ts/recipe/session/components/themes/styles.css:29]: Where is this style from? This doesn't really match our other buttons or uses our palette, etc.

File: lib/ts/recipe/session/components/themes/styles.css [hunk 35]
```diff
@@ -0,0 +1,51 @@
...
+
+@import "../../../../styles/styles.css";
+
+[data-supertokens~="center"] {
+    text-align: center;
+}
+
+[data-supertokens~="buttons"] {
+    display: flex;
+    gap: 20px;
```
Comment (porcellus) [inline lib/ts/recipe/session/components/themes/styles.css:24]: Does this prop do anything here?

File: lib/ts/recipe/session/index.ts [hunk 36]
```diff
@@ -128,11 +132,16 @@ const signOut = SessionAPIWrapper.signOut;
...
 const getInvalidClaimsFromResponse = SessionAPIWrapper.getInvalidClaimsFromResponse;
 const getClaimValue = SessionAPIWrapper.getClaimValue;
+const AccessDenied = SessionAPIWrapper.AccessDenied;
+const SessionComponentsOverrideProvider = SessionAPIWrapper.ComponentsOverrideProvider;
 
 export {
     useSessionContext,
     useClaimValue,
     SessionAuth,
+    AccessDeniedTheme,
```
Comment (porcellus) [inline lib/ts/recipe/session/index.ts:142]: This should be added to the class as well.

File: lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx [hunk 37]
```diff
@@ -1,40 +1,52 @@
...
     };
 
     const onBackButtonClicked = () => {
-        return SuperTokens.getInstanceOrThrow().redirectToAuth({
-            show: "signin",
-            history,
-            redirectBack: false,
-        });
+        try {
+            WindowHandlerReference.getReferenceOrThrow().windowHandler.getWindowUnsafe().history.back();
```
Comment (porcellus) [inline lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx:48]: I'd prefer if this used history (from `useHistoryCustom`) somehow.

File: lib/ts/recipe/session/recipe.tsx [hunk 38]
```diff
@@ -179,10 +179,7 @@ export default class Session extends RecipeModule<unknown, unknown, unknown, Nor
             });
 
             if (failureRedirectInfo.accessForbidden) {
-                await setLocalStorage(
-                    "supertokens-access-denial-info",
-                    JSON.stringify(failureRedirectInfo.failedClaim)
-                );
+                console.warn(failureRedirectInfo.failedClaim);
```
Comment (porcellus) [inline lib/ts/recipe/session/recipe.tsx:182]:
```suggestion
                console.warn({ message: "Redirecting to access denied screen", claimValidationError: failureRedirectInfo.failedClaim });
```

File: lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx (hunk 37 above)
Comment (porcellus) [reply lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx:48]: also, why does this have to be wrapped by try-catch?

**Review by porcellus — CHANGES_REQUESTED** [review]

**Review by elaxer — APPROVED** [review]

File: lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx [hunk 39]
```diff
@@ -0,0 +1,68 @@
...
+                redirectBack: false,
+            });
+        }
+    };
+
+    const onBackButtonClicked = () => {
+        if (props.history.back !== undefined) {
+            return props.history.back();
+        }
+        return props.history(-1);
```
Comment (porcellus) [inline lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx:36]: This needs to handle 3 cases:
- react-router-dom>6
- react-router-dom 5
- no react-router-dom

Did you test this? I don't think this works:
- history is undefined without react-router-dom (I think this is what you tried to handle)
- react-router-dom 5 history has `goBack`

File: lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx [hunk 40]
```diff
@@ -0,0 +1,68 @@
...
+import type { FC } from "react";
+
+const AccessDeniedScreen: FC<AccessDeniedThemeProps> = (props) => {
+    const userContext = useUserContext();
+    const t = useTranslation();
+    const onLogout = async () => {
+        try {
+            await props.recipe.signOut({ userContext });
+        } catch (error) {
+        } finally {
```
Comment (porcellus) [inline lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx:24]: Why is this wrapped in a try-catch?

**Review by porcellus — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/session/components/themes/styles.css [hunk 41]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+[data-supertokens~="buttonsGroup"] {
+    display: flex;
+    align-items: center;
+    justify-content: space-between;
+}
+
+[data-supertokens~="buttonBase"] {
+    font-size: 14px;
```
[the remark this answers is not in the document]
Comment (porcellus) [reply lib/ts/recipe/session/components/themes/styles.css]: There is still a font size not from  the palette.

File: lib/ts/styles/styles.css [hunk 42]
```diff
@@ -14,6 +14,7 @@
     --palette-textPrimary: 101, 101, 101;
     --palette-textLink: 0, 118, 255;
     --palette-buttonText: 255, 255, 255;
+    --palette-textDisabled: 128, 128, 128;
```
[the remark this answers is not in the document]
Comment (porcellus) [reply lib/ts/styles/styles.css]: Please add this to a changelog as well.

File: lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx [hunk 43]
```diff
@@ -17,21 +18,21 @@ import type { FC } from "react";
...
+        await SuperTokens.getInstanceOrThrow().redirectToAuth({
+            show: "signin",
+            redirectBack: false,
+        });
     };
 
     const onBackButtonClicked = () => {
-        if (props.history.back !== undefined) {
-            return props.history.back();
+        if (props.history === undefined) {
```
Comment (porcellus) [inline lib/ts/recipe/session/components/themes/accessDeniedScreenTheme/index.tsx:31]: - Please add an explanatory comment to say which branch handles which case.
- Add an end-to-end test for each branch please

File: lib/ts/recipe/session/sessionAuth.tsx [hunk 44]
```diff
@@ -175,17 +174,29 @@ const SessionAuth: React.FC<PropsWithChildren<SessionAuthProps>> = ({ children,
...
+                            history
+                        );
+                    }
+                    if (props.useDefaultAccessDeniedScreen !== false) {
+                        console.warn({
+                            message: "Access denied",
+                            claimValidationError: failureRedirectInfo.failedClaim,
+                        });
+                        return setContext({ ...toSetContext, accessDenied: true });
+                    }
```
Comment (porcellus) [inline lib/ts/recipe/session/sessionAuth.tsx:196]: This needs to happen even if `doRedirection` is false.

**Review by porcellus — CHANGES_REQUESTED** [review]
- Please merge the base-branch
- Update the changelog with more details

File: CHANGELOG.md [hunk 45]
```diff
@@ -7,6 +7,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [unreleased]
```
Comment (rishabhpoddar) [inline CHANGELOG.md:9]: update version to be 0.33.0

File: package.json [hunk 46]
```diff
@@ -83,7 +83,7 @@
     "peerDependencies": {
         "react": ">=16.8.0",
         "react-dom": ">=16.8.0",
-        "supertokens-web-js": "^0.5.0"
+        "supertokens-web-js": "github:supertokens/supertokens-web-js#0.6"
```
Comment (rishabhpoddar) [inline package.json:86]:
```suggestion
        "supertokens-web-js": "^0.6.0"
```

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]
